### ISSUE TITLE
Transformer model produces "." as output 

### ISSUE THREAD

**Issue by shredpub** [OpenNMT/OpenNMT-tf#83]
Hello,
I've been trying to train a transformer model using the latest code version.
I tried using tensorflow 1.6 (had to convert the code to python3 for that) as well as tensorflow 1.5.
However, I get only "." as the output on testing the model using the toy-ende dataset provided in the distribution. I use the transformer.py and the transformer1gpu.yml as the config files. 
On the other hand, training a vanilla RNN model (with nmt_small.py and opennmtdefaults.yml) works fine and produces legit outputs. Any lead on this?

**Comment by guillaumekln** [comment]
Hello,

Sorry for that, I introduced a regression in https://github.com/OpenNMT/OpenNMT-tf/commit/f59f011a8cc84b025b8d260503968cf33f650bdc that I just fixed a few minutes ago in https://github.com/OpenNMT/OpenNMT-tf/commit/40ce318368a275f98f95662274f1f5f0948a44e5.

Could you update the code and restart the training?

Thanks.

**Comment by shredpub** [comment]
Thanks for the response. I shall restart the training. Before that, could you let me know if there's a config file for Transformer with multi-gpu training? I see only one for single gpu.

**Comment by guillaumekln** [comment]
You can use the same file but apply the recommendation of line 10:

https://github.com/OpenNMT/OpenNMT-tf/blob/047ac49aa77c1dfaa805755e40e3245f8960eab3/config/transformer_1gpu.yml#L10-L11

then set the number of GPUs with the option `--num_gpus`.

Also see the WMT English-German training scripts: https://github.com/OpenNMT/OpenNMT-tf/tree/master/scripts/wmt

**Comment by shredpub** [comment]
Thanks! Which tensorflow version do you recommend to be used?

**Comment by guillaumekln** [comment]
This is not relevant to the current issue (you can use the [Gitter channel](https://gitter.im/OpenNMT/OpenNMT-tf) for such questions). To answer it anyway, the code on `master` should be compatible with all TensorFlow versions superior or equal to 1.4. If not, open an issue!

**Comment by shredpub** [comment]
Thanks! Closing this issue for now.

**Comment by shredpub** [comment]
Update : It still doesn't work with the updated code. I get the following error on testing : 
_INFO:tensorflow:Restoring parameters from transformerEnIt/model.ckpt-20001
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Aborted_

**Comment by guillaumekln** [comment]
Looks like there is not enough memory. Can you check?

**Comment by shredpub** [comment]
Well I ran on 12 GB GPU memory on a not so large dataset. It hangs up for quite some time on testing, after which it throws the above mentioned error. 
I'm testing again on toy-ende with higher GPU memory just for confirming.

**Comment by shredpub** [comment]
Reproduced on toy-ende. Freezes on testing (even on 36 GB GPU memory) - both for single gpu as well as multi gpu. Using standard config files transformer_1gpu and transformer.py without any modification.

**Comment by guillaumekln** [comment]
What is the TensorFlow you are using? You mentioned both 1.5 and 1.6 in the first comment. It seems others encounter such issues with 1.5.

**Comment by shredpub** [comment]
I am using 1.5.

**Comment by guillaumekln** [comment]
Can you try with 1.6 instead?

**Comment by shredpub** [comment]
Works with 1.6, thanks. Closing the issue.

**Comment by AnubhavSi** [comment]
Can you please explain why 1.5 is not working?
I am training it on ML-engine with the same configuration with 4 gpu and after submitting the job, it's throwing this error. 

The replica master 0 ran out-of-memory and exited with a non-zero status of 9(SIGKILL). To find out more about why your job exited please check the logs: https://console.cloud.google.com/logs/viewer?project=1002083926433&resource=ml_job%2Fjob_id%2Fnmt_v1_27&advancedFilter=resource.type%3D%22ml_job%22%0Aresource.labels.job_id%3D%22nmt_v1_27%22

**Comment by guillaumekln** [comment]
Out of memory is a common issue but most likely not related to TensorFlow 1.5 in particular. Please open a new issue if you feel like there is bug. Otherwise, you should probably reduce the training batch size, or train a smaller model to fit in the GPU memory available to you.
